server/src/main/java/mineclone/server/world/block/signal/wire/WireHandler.java

```diff
@@ -0,0 +1,1301 @@
+	 */
+	private void searchRoot(WireNode wire) {
+		int iBackupFlowDir;
+
+		if (wire.connections.iFlowDir < 0) {
+			iBackupFlowDir = 0;
+		} else {
+			iBackupFlowDir = wire.connections.iFlowDir;
+		}
+
+		search(wire, true, iBackupFlowDir);
+	}
+
+	/**
+	 * Queue the given wire for the breadth-first search and set a backup flow
+	 * direction.
+	 */
+	private void search(WireNode wire, boolean root, int iBackupFlowDir) {
+		search.offer(wire);
+
+		wire.root = root;
+		wire.searched = true;
+		// Normally the flow is not set until the power level is updated. However,
+		// in networks with multiple power sources the update order between them
+		// depends on which was discovered first. To make this less prone to
+		// directionality, each wire node is given a 'backup' flow. For roots, this
+		// is the determined flow of their connections. For non-roots this is the
+		// direction from which they were discovered.
+		wire.iFlowDir = iBackupFlowDir;
+	}
+
+	private void tryUpdate() {
+		if (!search.isEmpty()) {
+			update();
+		}
+		if (!updating) {
+			nodes.clear();
+			nodeCount = 0;
+		}
+	}
+
+	/**
+	 * Update the network and neighboring blocks. This is done in 3 steps.
+	 * 
+	 * <p>
+	 * <b>1. Search through the network</b>
+	 * <br>
+	 * Conduct a breadth-first search around the roots to find wires that are in an
+	 * invalid state and need power changes.
+	 * 
+	 * <p>
+	 * <b>2. Depower the network</b>
+	 * <br>
+	 * Depower all wires in the network. This allows power to be spread most
+	 * efficiently.
+	 * 
+	 * <p>
+	 * <b>3. Power the network</b>
+	 * <br>
+	 * Work through the update queue, setting the new power level of each wire and
+	 * updating neighboring blocks. After a wire has updated its power level, it
+	 * will emit shape updates and queue updates for neighboring wires and blocks.
+	 */
+	private void update() {
+		// The profiler keeps track of how long various parts of the algorithm take.
+		// It is only here for debugging purposes, and is commented out in production.
+//		Profiler profiler = AlternateSignalMod.createProfiler();
+//		profiler.start();
+
+		// Search through the network for wires that need power changes. This includes
+		// the roots as well as any wires that will be affected by power changes to
+		// those roots.
+//		profiler.push("search network");
+		searchNetwork();
+
+		// Depower all the wires in the network.
+//		profiler.swap("depower network");
+		depowerNetwork();
+
+		// Bring each wire up to its new power level and update neighboring blocks.
+//		profiler.swap("power network");
+		try {
+			powerNetwork();
+		} catch (Throwable t) {
+			// If anything goes wrong while carrying out power changes, this field must
+			// be reset to 'false', or the wire handler will be locked out of carrying
+			// out power changes until the world is reloaded.
+			updating = false;
+
+			throw t;
+//		} finally {
+//			profiler.pop();
+//			profiler.end();
+		}
```

Comment:
For the sake of readability I think it makes sense to move the updating flag outside of `#powerNetwork()`. I suggest the following instead:
```suggestion
		if (!updating) {
			updating = true;
			try {
				powerNetwork();
			} finally {
				// If anything goes wrong while carrying out power changes, this field must
				// be reset to 'false', or the wire handler will be locked out of carrying
				// out power changes until the world is reloaded.
				updating = false;
			}
		}
```
If the above suggestion is implemented, remember to remove it from `#powerNetwork()`.